include/rocksdb/utilities/cache_dump_load.h

```diff
@@ -0,0 +1,77 @@
+//  Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <set>
+
+#include "rocksdb/cache.h"
+#include "rocksdb/env.h"
+#include "rocksdb/io_status.h"
+#include "rocksdb/secondary_cache.h"
+#include "rocksdb/table.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+class CacheDumpWriter {
+ public:
+  virtual ~CacheDumpWriter() = default;
+
+  virtual IOStatus Write(const Slice& data) = 0;
+  virtual IOStatus Close() = 0;
+  virtual uint64_t GetFileSize() = 0;
+};
+
+class CacheDumpReader {
+ public:
+  virtual ~CacheDumpReader() = default;
+  virtual IOStatus Read(size_t len, std::string* data) = 0;
+  virtual size_t GetOffset() const = 0;
+  virtual IOStatus Close() = 0;
+};
+
+// CacheDumpOptions is used to control DumpCache()
+struct CacheDumpOptions {
+  // The env pointer
+  Env* env;
+  // The prefix set, which filter out the cache keys that does not match
+  std::set<std::string> prefix_set;
+};
+
+class CacheDumper {
+ public:
+  virtual ~CacheDumper() = default;
+  virtual Status Prepare() = 0;
+  virtual IOStatus Run() = 0;
+};
+
+class CacheDumpedLoader {
+ public:
+  virtual ~CacheDumpedLoader() = default;
+  virtual Status Prepare() = 0;
+  virtual IOStatus Run() = 0;
+};
+
+Status NewDefaultCacheDumpWriter(Env* env, const EnvOptions& env_options,
+                                 const std::string& file_name,
+                                 std::unique_ptr<CacheDumpWriter>* writer);
+
+Status NewDefaultCacheDumpReader(Env* env, const EnvOptions& env_options,
+                                 const std::string& file_name,
+                                 std::unique_ptr<CacheDumpReader>* reader);
+
+Status NewDefaultCacheDumper(const CacheDumpOptions& dump_options,
+                             const std::shared_ptr<Cache>& cache,
```

Comment:
The Cache to dump is more general than the filter to apply to entries (currently `prefix_set`). So it doesn't make sense to me for the more general one to be an input and the less general one part of an Options object.

Reply:
Agree. It might be better to have CacheDumper::SetDumpFilter(std::vector<DB*> db_list)?